modules/p2/ding_interests/ding_interests.module

```diff
@@ -0,0 +1,652 @@
+        $variables['content']['ding_interest_no_interest_button'] = array(
+          '#type' => 'markup',
+          '#markup' => t('Already added to !list', array('!list' => l(t('interests'), 'list/' . $list->id))),
+          '#weight' => 50,
+        );
+      }
+    }
+  }
+  elseif ($view_mode == 'teaser') {
+    drupal_add_css(drupal_get_path('module', 'ding_interests') . '/ding_interests.css');
+    $variables['classes_array'][] = 'no-content';
+  }
+}
+
+/**
+ * Ajax callback for the add to list button.
+ *
+ * @ajax boolean
+ *   Indicates if this call is running through Ajax.
+ *
+ * @term object taxonomy_term
+ *   The taxonomy term being added.
+ *
+ * @account object user
+ *   The user to act on.
+ *
+ * @return array
+ *   if ajax is used an array of ajax commands is returned,
+ *   else an empty string is returned.
+ */
+function ding_interests_add_to_list_callback($ajax, $term_id, $account, $rating = 5) {
+  if ($ajax) {
+    $term = ding_list_get_taxonomy_term($term_id);
+
+    // Add to interests list.
+    $added_result = ding_interests_add_to_list($term, $account, $rating);
+    $list = ding_list_get_list(DING_INTERESTS_LIST, $account);
+    // Get ajax button.
+    $link = ($rating == DING_INTERESTS_INTERESTING) ? l(t('interests'), 'list/' . $list->id) : l(t('not interested'), 'list/' . $list->id);
+    $output = '<div class="response">' . t('Added to !list', array('!list' => $link)) . '</div>';
+
+    if (!isset($_GET['frombutton'])) {
+      // Set command.
+      $identifier = '#taxonomy-term-' . $term->tid . ' .add-to-list';
+      $commands[] = ajax_command_replace($identifier, $output);
+    }
+    else {
+      if (!$added_result) {
+        $response = t('Already added to !name', array('!name' => l($list->title, 'list/' . $list->id)));
+        $commands[] = ajax_command_ding_popup('ding_list', t('List'), '<div class="ding-list-message">' . $response . '</div>');
+      }
+      else {
+        $response = t('Added to !name', array('!name' => l($list->title, 'list/' . $list->id)));
+        $commands[] = ajax_command_ding_popup('ding_list', t('List'), '<div class="ding-list-message">' . $response . '</div>');
+      }
+    }
+
+    if ($rating > DING_INTERESTS_NOT_INTERESTING) {
+      $commands[] = ding_serendipity_command_refresh(array('interest' => $term->tid, 'plugins' => array('ding_interests_materials')));
+    }
+    // Return output.
+    ajax_deliver($commands);
+  }
+  else {
+    // Ajax required.
+    return '';
+  }
+}
+
+/**
+ * Ajax callback for the remove from list button.
+ *
+ * @ajax boolean
+ *   Indicates if this call is running through Ajax.
+ *
+ * @term object taxonomy_term
+ *   The taxonomy term being added.
+ *
+ * @account object user
+ *   The user to act on.
+ *
+ * @return array
+ *   if ajax is used an array of ajax commands is returned,
+ *   else an empty string is returned.
+ */
+function ding_interests_remove_from_list_callback($ajax, $term, $account) {
+  if ($ajax) {
```

Comment:
There does not seem to be any check that $account is either the current user or one with sufficient privileges. In practices this means that any user will be able to unmark a term as interesting for another user by changing the url.